### PR TITLE
perf: Store indent descriptors in a plain array

### PR DESCRIPTION
**Pull request by fasttime**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Improves performance of the `indent` rule. Makes the build faster.

#### What changes did you make? (Give an overview)

This PR changes the underlying implementation of `BinarySearchTree` in the `indent` rule to a plain array. The name `BinarySearchTree` is changed to `IndexMap` to avoid confusion now that a tree-like structure is no longer used. I also renamed a few other identifiers for clarity.

##### How it works

The `BinarySearchTree`/`IndexMap` class is used by the `indent` rule to map runtime information (*descriptors*) to indices in the source code. You can think of it as a mapping of integers in a known interval to arbitrary objects.
The performance of this map is determined by the time consumed by the three methods that operate on it.
* Add a mapping (`insert`)
* Remove all mappings in a specified index range (`deleteRange`)
* Search backwards for the first mapping, starting at a specified index (`findLe`/`findLastNotAfter`)

In the previous implementation, a red-black tree (provided by [this package](https://www.npmjs.com/package/js-sdsl)) was used to store the mappings, whereas the new implementation uses a constant-size array. With this new implementation, adding a mapping only requires setting an element of the array. Deleting a range takes a single call to the built-in method `fill`. Obviously, both operations are much more efficient compared to their red-black tree counterparts.
As for the backward search, it now requires iterating indices one by one until a mapping is found. The exact number of iterations depends on the length and arrangement of tokens and spaces in the source file. While this value can grow up to the number of characters in the file + 1 in the worst case, it will typically not depend on the file size, but rather on the coding style. For reference, when linting the `eslint` repo, the average search takes about 47 iterations, and performs still much faster than the same search with the old implementation.

##### Measuring performance

I run `npm run lint` after setting the environment variable `TIMING=1`, and compared the results before and after the change.

##### Results

I tested these changes on MacOS and Windows with different versions of Node.js, with similar results. After the changes, the `indent` rule runs about 40% faster than before.

Example run:

**Before**

```text
Rule                                          | Time (ms) | Relative
:---------------------------------------------|----------:|--------:
indent                                        |  3228.423 |    17.8%
```

**After**

```text
Rule                             | Time (ms) | Relative
:--------------------------------|----------:|--------:
indent                           |  1656.434 |    10.0%
```

### CODE REVIEW

**Review by aladdin-add — APPROVED** [review]
LGTM. Awesome job! 👍

waiting for a few more review before merging.

**Review by nzakas — APPROVED** [review]
Wow, that's a really impressive speed up. LGTM. We'll just need to wait until after Monday to merge.

File: lib/rules/indent.js
Comment (mdjermanovic) [inline lib/rules/indent.js:155]:
```suggestion
     * @returns {*|undefined} The value of the found entry, or undefined if no such entry exists.
```

This should be `*` because it returns value, not key?

Comment (fasttime) [reply lib/rules/indent.js]: Thanks! Yes, that's the value.

**Pull request merged**
